tests/treevec_test.py

```diff
@@ -0,0 +1,534 @@
+
+    expected = {'a': 2.0, 'b': -2.0}
+    self.assertAllClose(actual, expected, check_dtypes=True)
+
+  def test_vmap(self):
+
+    def square_scalar(x):
+      assert x.shape == ()
+      return jnp.square(x)
+
+    @tree_vectorize
+    def f(x):
+      return jax.vmap(square_scalar)(x)
+
+    tree = {'x': 2.0, 'y': np.array([3.0, 4.0])}
+    expected = {'x': 4.0, 'y': np.array([9.0, 16.0])}
+    actual = f(tree)
+    self.assertAllClose(actual, expected, check_dtypes=True)
+
+  def test_vmap_tree_call(self):
+
+    def g(x):
+      assert x['a'].shape == ()
+      return {'b': x['a']}
+
+    @tree_vectorize
+    def f(g, x):
+      return jax.vmap(g)(x)
+
+    tree = {'a': jnp.arange(3.0)}
+    expected = {'b': jnp.arange(3.0)}
+    actual = f(g, tree)
+    self.assertAllClose(actual, expected, check_dtypes=True)
+
+  def test_vmap_tree_call_in_axes(self):
+
+    def g(x, y):
+      assert x['a'].shape == ()
+      assert y['a'].shape == (3,)
+      return {'b': x['a'] + y['a']}
+
+    @tree_vectorize
+    def f(g, x, y):
+      return jax.vmap(g, in_axes=(0, None))(x, y)
+
+    trees = ({'a': jnp.arange(3.0)}, {'a': 10 * jnp.ones(3)})
+    expected = {'b': jnp.array([[10.0, 10, 10], [11, 11, 11], [12, 12, 12]])}
+    actual = f(g, *trees)
+    self.assertAllClose(actual, expected, check_dtypes=True)
+
+  def test_jvp(self):
+    @tree_vectorize
+    def f(x, y):
+      return jax.jvp(lambda x: 0.5 * x ** 2, (x,), (y,))
+
+    primal = {'x': 1.0, 'y': 2.0}
+    tangent = {'x': 3.0, 'y': 4.0}
+    actual = f(primal, tangent)
+    expected = ({'x': 0.5, 'y': 2.0}, {'x': 3.0, 'y': 8.0})
+    self.assertAllClose(actual, expected, check_dtypes=True)
+
+  def test_jvp_tree_call(self):
+
+    def g(x):
+      print("g(x) called with", x)
+      return {'b': 0.5 * x['a'] ** 2}
+
+    @tree_vectorize
+    def f(g, x, y):
+      z, dz = jax.jvp(g, (x,), (y,))
+      return z, dz
+
+    actual = f(g, {'a': 2.0}, {'a': 3.0})
+    expected = ({'b': 2.0}, {'b': 6.0})
+    self.assertTreeEqual(actual, expected, check_dtypes=True)
+
+  @pytest.mark.xfail
+  def test_grad(self):
+    def g(x):
+      return x['a'] + 2 * x['b'] + x['c'] ** 2
+
+    @tree_vectorize
+    def f(g, x):
+      return jax.grad(g)(x)
+
+    actual = f(g, {'a': 0.0, 'b': 0.0, 'c': -1.0})
+    expected = {'a': 1.0, 'b': 2.0, 'c': -2.0}
+    self.assertTreeEqual(actual, expected, check_dtypes=True)
+
+  def _jac_apply_argnums(self, transform, fun, argnums=0):
+    def jacfun(*args):
+      f_partial, dyn_args = api.argnums_partial(
+          lu.wrap_init(fun), argnums, args)
+      result = transform(f_partial.call_wrapped, dyn_args)
+      return result[0] if isinstance(argnums, int) else result
+    return jacfun
+
+  @pytest.mark.xfail
+  def test_jacfwd(self):
+
+    @tree_vectorize
+    def _jacfwd(f, x):
+      pushfwd = partial(jax.jvp, f, (x,))
+      basis = jnp.eye(x.size, dtype=x.dtype)
+      y, jac = jax.vmap(pushfwd, out_axes=(None, 1))((basis,))
```

Comment:
I think this is very close to working, except we don't handle `basis` properly yet because it has a different tree structure. (This works in other cases, but not for transformed tree_call functions yet.)